SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/CrossServerRPCHandler.h"
+
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
+
+using namespace SpatialGDK;
+
+CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor)
+	: Coordinator(InCoordinator)
+	, RPCExecutor(MoveTemp(InRPCExecutor))
+{
+}
+
+void CrossServerRPCHandler::ProcessOps(const float TimeAdvancedS, const TArray<Worker_Op>& WorkerMessages)
+{
+	for (auto& Op : WorkerMessages)
+	{
+		if (Op.op_type == WORKER_OP_TYPE_COMMAND_REQUEST
+			&& Op.op.command_request.request.command_index == SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID
+			&& Op.op.command_request.request.component_id == SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID)
+		{
+			HandleWorkerOp(Op);
+		}
+	}
+
+	ProcessPendingCommandOps();
+}
+
+void CrossServerRPCHandler::ProcessPendingCommandOps()
+{
+	TArray<Worker_EntityId> EmptyRPCQueues;
+	for (auto& CrossServerRPCs : QueuedCrossServerRPCs)
+	{
+		int32 ProcessedRPCs = 0;
+		for (auto& Command : CrossServerRPCs.Value)
+		{
+			if (!TryExecuteCommandRequest(Command))
+			{
+				break;
+			}
+
+			RPCGuidsInFlight.Remove(Command.Payload.UniqueId);
+			++ProcessedRPCs;
+		}
+
+		CrossServerRPCs.Value.RemoveAt(0, ProcessedRPCs);
+		if (CrossServerRPCs.Value.Num() == 0)
+		{
+			EmptyRPCQueues.Add(CrossServerRPCs.Key);
```

Comment:
You do not need an extra array for empty queues if you use the Map's Iterator::RemoveCurrent()

Reply:
ah cool. I wasn't sure if RemoveCurrent works for this. will change it

Reply:
Is that documented anywhere. I spent a while trying to find out if that was actually valid or not and stopped before getting onto reading through the entire implementation to see. Although I presume it replaces the element with a sentinel.

Reply:
If RemoveCurrent on an iterator made it unusable, it would defeat the purpose of having the method in the first place.